LLM tool support for python prism

chitacan opened the pull request:
This pull request introduces updates to the Lux framework, focusing on enhancing Python integration and improving type handling between Elixir and Python. Now we can specify `{:python, "path/to/prism.py"}` in the `Lux.Agent`'s `prisms`.

```elixir
defmodule MyApp.Agent do
  use Lux.Agent,
    description: "An assistant that greats you",
    prisms: [
      {:python, "path/to/prism.py"}, # Python Prism 🐍
      MyApp.Prisms.DataAnalysis      # Other Prism
    ]
end
```

Later on, the agent can extract metadata from the python prism and use it as a tool.

```elixir
MyApp.Agent.chat(agent, "Hello my name is John")
# => "Hello, John!"
```

## How it works?

When we call `Lux.Prism.view("path/to/prism.py")`, I added a some private function to dynamically create modules for Python prisms. I chose this method because it is easy to integrate the LLM tool calling if there is an elixir side module. To achieve this, when defining a Python Prism, we have to follow the convention of naming it the same as the elixir module convention. (see)

```elixir
Lux.Prism.view("path/to/prism.py")
# Simple.Prism module will be defined
Simple.Prism.handler(%{name: "John"}, %{})
```

## Improvements in tuple type handling in python

Enhanced encoding and decoding functions in `lux/priv/python/lux/eval.py` to properly handle tuples. The `eval` function now properly convert python tuple to elixir tuple.

AS-IS:
```elixir
Lux.Python.eval("(1,2)")
# {:ok, [1, 2]}
```

TO-BE:
```elixir
Lux.Python.eval("(1,2)")
# {:ok, {1, 2}}
```

chitacan commented on lux/guides/prisms.livemd:
```diff
@@ -549,6 +549,68 @@ MyApp.Prisms.CryptoAddressValidator.run(%{
+class SimplePrism(Prism):
+    id='aaace2b1-8089-4d4e-b68e-2ce904da12f0'
+    # Make sure that you follow the elixir's naming convention.
+    name='Simple.Prism'
```
The elixir's module name for python prism came from its `name`. That's why the python prism's name should follow elixir's naming convention.

In this case, after we call `view`, `Simple.Prism` will be defined dynamically.

I'll update the doc to clarify this process.

rparcus replied:
The name thing is intended to be the name we pass to the LLM when it does tool calling, not necessarily tied to an elixir module name. Could you add the module_name attribute and if nil we could fallback to the ClassName and then to `name`?

chitacan replied:
Yes, for python prism, it would be good to declare the `module_name` attribute separately. I will update this part.

For reference, in the current implementation, if the name of prism is not the module name, the tool calling is implemented to fail.

https://github.com/Spectral-Finance/lux/blob/b9c2cca11e5646d98ccbcc49bcfd3230594afa53/lux/lib/lux/llm/open_ai.ex#L319-L334

rparcus replied:
>For reference, in the current implementation, if the name of prism is not the module name, the tool calling is implemented to fail.

I don't understand this, can you show me where are we using Prism.name to call the module?

I think that in elixir Prism names should be settable to any string, and if missing then we fallback to a string version of the module name 
![image](https://github.com/user-attachments/assets/7ec26ccf-aab2-4e3d-9f63-bb78da3cae03)

chitacan replied:
When we registering a tool from prism, we derive tool's name based on prism's name:

https://github.com/Spectral-Finance/lux/blob/b9c2cca11e5646d98ccbcc49bcfd3230594afa53/lux/lib/lux/llm/open_ai.ex#L207-L217

Later, when LLM response with tool calling response, we convert tool's name to elixir module format and call `handler/2`:

https://github.com/Spectral-Finance/lux/blob/b9c2cca11e5646d98ccbcc49bcfd3230594afa53/lux/lib/lux/llm/open_ai.ex#L297-L317

Currently, using the following names in prism may result in failure during llm tool registration and calling.

* `Simple Prism`: OpenAI function name only allows `[a-zA-Z0-9_-]`.
* `Simple-Prism`: The function call fails if the actual declared module name is different.

rparcus replied:
Ok, got it. Thanks for the refresher!

rparcus replied:
I think we might be able to improve this further (another time though)

chitacan commented on lux/lib/lux/prism.ex:
```diff
@@ -138,11 +133,12 @@ defmodule Lux.Prism do
-      }
-    )
+    with {:ok, prism} <- Lux.Python.eval(path, variables: %{__lux_function__: :view}),
+         :ok <- define_module_if_not_exists(prism, path) do
```
Well, I'm not sure about that too. After thinking about it for a moment, I think the `code_server` app might be doing dynamic module compilation. For more details, let me take a look at the source code of the `Module.create/2` function.

chitacan replied:
I've examined the source code of `Module.create/2`. Looks like there is no special processing to process them in parallel. Only successfully compiled module will be added to the `code_server`. If multiple processes declare modules with the same name at the same time, they may be overwritten by the last module declared.

For our implementation, I skip the module creation process if there's a same name in elixir runtime.

chitacan replied:
Modified to output a warning log when a module other than prism is already declared with the same name. [957503c](https://github.com/Spectral-Finance/lux/pull/139/commits/957503c966bece428e7412fa5a5f8ea2795f1e91)

chitacan commented on lux/guides/prisms.livemd:
````diff
@@ -549,6 +549,68 @@ MyApp.Prisms.CryptoAddressValidator.run(%{
+    ]
+end
+```
+
````
Link added [b9c2cca](https://github.com/Spectral-Finance/lux/pull/139/commits/b9c2cca11e5646d98ccbcc49bcfd3230594afa53)

chitacan commented:
@rparcus I've updated the documentation based on your feedback. Could you please take another look?

rparcus commented on lux/guides/prisms.livemd:
```diff
@@ -576,8 +578,9 @@ class SimplePrism(Prism):
-After loading the prism, you can use it like any other Elixir prism:
+Only the packages listed in [`pyproject.toml`](../priv/python/pyproject.toml) are available to your Python prisms.
+While you access the prism with `Lux.Prism.view/1`, `Simple.Prism` module is dynamically defined and available in Elixir runtime.
+(The module name is derived from the `name` attribute of the Python prism class.)
```
So, let's not do this.

chitacan commented:
@rparcus Hey, I updated the python prism base class and elixir module, so user don't need to decide the name for elixir module convention. I added `module_name` field in prism struct, it will automatically determine appropriate value for it.

* for python prism: use python class name
* for elixir prism: use elixir module name

Always, thanks for the great feedback!!

rparcus reviewed (approved):
Amazing work @chitacan!
